core/src/main/scala/kafka/log/Log.scala

```diff
@@ -544,6 +574,97 @@ class Log(val dir: File,
   }
 
   /**
+   * Based on the given timestamp, get a sequence of specified number of offsets in descending order.
+   * The first offsets returned in the offset sequence is the offset of the first message whose timestamp is greater
+   * or equals to the given timestamp. The rest of the offsets are the closest but smaller offsets in the time index.
+   *
+   * `NOTE:` To maintain the behavior of OffsetRequest V0, if the given timestamp is smaller or equals to the earliest
+   * timestamp in the log, an empty sequence is returned.
+   *
+   * @param timestamp The given timestamp for offset fetching.
+   * @param maxNumOffsets the maximum number of offsets to return.
+   * @return A sequence of offsets in descending order.
+   */
+  def fetchOffsetsByTimestamp(timestamp: Long, maxNumOffsets: Int): Seq[Long] = {
```

Comment:
The current getOffsetBefore() api is a bit awkward to use since it's tightly coupled with how we use the last modified time in each log segment. For example, it doesn't seem to make sense to return a sequence of offsets now that we have message level timestamp. So, it's probably simpler and better to leave the current implementation of getOffsetBefore() as it is. In Log, LogSegment, and TimeIndex, we expose a simple api that takes a timestamp and returns a single offset whose timestamp is >= than the input. In the future, we can design a new request that exposes this capability to the client.